internal/coord/behaviour.go

```diff
@@ -146,3 +148,48 @@ func (w *Waiter[E]) Close() {
 func (w *Waiter[E]) Chan() <-chan WaiterEvent[E] {
 	return w.pending
 }
+
+// A QueryMonitor receives event notifications on the progress of a query
+type QueryMonitor[E TerminalQueryEvent] interface {
+	// NotifyProgressed returns a channel that can be used to send notification that a
+	// query has made progress. If the notification cannot be sent then it will be
+	// queued and retried at a later time. If the query completes before the progress
+	// notification can be sent the notification will be discarded.
+	NotifyProgressed() chan<- CtxEvent[*EventQueryProgressed]
+
+	// NotifyFinished returns a channel that can be used to send the notification that a
+	// query has completed. It is up to the implemention to ensure that the channel has enough
+	// capacity to receive the single notification.
+	// The sender must close all other QueryNotifier channels before sending on the NotifyFinished channel.
+	// The sender may attempt to drain any pending notifications before closing the other channels.
+	// The NotifyFinished channel will be closed once the sender has attempted to send the Finished notification.
+	NotifyFinished() chan<- CtxEvent[E]
+}
```

Comment:
Given the current usage of the `QueryMonitor` I think it would be a nicer API if these were just regular methods that accepted `CtxEvent[*EventQueryProgressed]` and `CtxEvent[E]` events. I can only see it used down below as:

```go
func (w *queryNotifier[E]) TryNotifyProgressed(ctx context.Context, ev *EventQueryProgressed) bool {
	if w.stopping {
		return false
	}
	ce := CtxEvent[*EventQueryProgressed]{Ctx: ctx, Event: ev}
	select {
	case w.monitor.NotifyProgressed() <- ce:
		return true
	default:
		w.pending = append(w.pending, ce)
		return false
	}
}

func (w *queryNotifier[E]) NotifyFinished(ctx context.Context, ev E) {
	w.stopping = true
	w.DrainPending()
	close(w.monitor.NotifyProgressed())

	select {
	case w.monitor.NotifyFinished() <- CtxEvent[E]{Ctx: ctx, Event: ev}:
	default:
	}
	close(w.monitor.NotifyFinished())
}
```

This requires users of the types that implement this interface to deal with quite some internal details.

Alternative suggestion:

```go
// A QueryMonitor receives event notifications on the progress of a query
type QueryMonitor[E TerminalQueryEvent] interface {
	NotifyProgressed(e CtxEvent[*EventQueryProgressed]) bool // indicating successful notification
	NotifyFinished(e CtxEvent[E])
}
```

closing the specific channels could happen inside the type that implements that interface.